Added simple game updating, list view in addition to grid view

mdgomes opened the pull request:
* Added a simple game updating, probably only works for linux (implements #1)
* Added a list view in addition to the grid view
* Added more information to the game class
* Update validation for linux games (takes into account current installed version versus the latest installer version)

Future work:
* gametile and gamerow can probably have a common parent

sharkwouter commented:
Wow, this is massive. I might need some time to review this and figure out how it works. I really appreciate the PR, thanks!

One point of critique for now, is that it is preferable to make a PR for each feature. That way it's a bit easier to review. But for now I'll just test this one and let you know my taughts.

mdgomes commented:
Yes, I got a bit carried away, sorry.

This can also work for issue #106 (list view has a bit more details, full name and category/genre)

mdgomes commented:
Also, to clarify a bit how the update is handled.

- When we are pulling the user's game list the updates flag is checked (matches what you get on gog.com) and assigned to the game class
- For installed games, product info is pulled rom the API and the installed version (from the gameinfo file) is compared to the latest installer version, if it is the same, the game.updates flag is overriden to not show updates (i.e. game.updates = 0 instead of 1)
- If the game is not installed the gtk-refresh icon is shown overlayed on the image
- if the game is installed a custom image is shown instead (more visible)

I consolidated some of the Library stuff on a helper class (seeing I split into two classes). One of the things that changed is that the filter method supports filtering games by more stuff (categories, platforms and tags).

I think I added all strings to the PO file but I might have missed something.

sharkwouter commented:
After cloning this and running it, I see a couple of issues.

The wine games I have installed do not show their thumbnails anymore:
![Screenshot from 2020-05-08 18-06-23](https://user-images.githubusercontent.com/5042659/81424697-a504ed80-9156-11ea-851c-1994165788cb.png)

When I tried updating Crystal Caves, I got the following error:
```
Executing integrity check for /home/wouter/.cache/minigalaxy/update/Crystal Caves
Verifying archive integrity... MD5 checksums are OK. All good.
warning [/home/wouter/.cache/minigalaxy/update/Crystal Caves]:  811603 extra bytes at beginning or within zipfile
  (attempting to process anyway)
Exception in thread Thread-535:
Traceback (most recent call last):
  File "/usr/lib/python3.7/threading.py", line 917, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.7/threading.py", line 865, in run
    self._target(*self._args, **self._kwargs)
  File "/home/wouter/Sources/minigalaxy/minigalaxy/download.py", line 25, in finish
    self.__finish_func()
  File "/home/wouter/Sources/minigalaxy/minigalaxy/ui/gametile.py", line 303, in __update
    install_game(self.game, self.update_path, parent_window=self.parent)
  File "/home/wouter/Sources/minigalaxy/minigalaxy/installer.py", line 47, in install_game
    shutil.copytree(tmp_noarch_dir, game.install_dir,dirs_exist_ok=True)
TypeError: copytree() got an unexpected keyword argument 'dirs_exist_ok'
```

I've also noticed that if you switch view while you have the installed switch at the top set to on, it will show all games anyway.

![Screenshot from 2020-05-08 18-09-24](https://user-images.githubusercontent.com/5042659/81424996-12188300-9157-11ea-81d0-de517c143a2e.png)
Btw, what do the white arrows mean?

mdgomes commented:
About the withe arrows, it's for games that have been updated since last you went into the gog.com interface (see comment above). I can remove the non-installed arrows and use the white arrow for the installed games (gets rid of the png image also)

The exception might be a flag available since python 3.8, let me check and update the PR if that is the case.

mdgomes commented:
Okay, fixed the issue with the installed filter not sticking after change, also fixed the cause for widows icons not showing and added a copytree function that does the same as 3.8 version. Try now.

Regarding the update icon, take your pick and I'll update the PR it's fine for me either way.

sharkwouter commented:
I'm really sorry for not looking into this in a while. I'll make time to do a formal review this weekend.

sharkwouter commented:
I see saving which view is used it works now.

I still see the following issues:
- If showing Windows games is enabled, it shows there are updates available for all installed Windows games.
- When switching between different views, the API is used again. Since the library has which games are installed and available, this shouldn't happen. It makes switching take too long.
- I did some testing with the update system. For me it shows that certain games always have and update available, even if you've already run the update. So there is something wrong with the check to see if there is an update available. Could you describe how you did this?

Besides this, I don't really like the white errors on games which aren't installed. Could they be left out?

sharkwouter commented on data/po/minigalaxy.pot:
```diff
@@ -336,10 +340,21 @@ msgstr ""
+msgid "Download Update"
+msgstr ""
+
+msgid "Install Update"
```
Do we need both download and install strings for update?

sharkwouter commented on data/ui/application.ui:
```diff
@@ -71,8 +71,8 @@
     <property name="window_position">center</property>
-    <property name="default_width">1400</property>
-    <property name="default_height">800</property>
+    <property name="default_width">1024</property>
```
This is where the default window size is changed. I'd prefer this to not be the case.

sharkwouter commented on minigalaxy/ui/list_library.py:
```diff
@@ -0,0 +1,112 @@
+        self.search_string = search_string
+        self.offline = False
+        self.library = Library(self.api)
+        self.games = []
```
Why not use the games list from the library object?

sharkwouter commented:
Thanks for updating! I'll take a look at it on the weekend

sharkwouter commented:
Sorry about taking ages to review this. The last while has been a bit hectic. I moved country, handed in my thesis and I'm trying to revive my social life. 

I did now review this now, but I'm seeing quite a few issues with it :(

Here are some things I noticed in my short test:

- It takes like a minute before any games show up in my list after launching.
- If you have the same game installed twice with different names, the image won't be loaded for the second one. This is a little bit broken in the latest release as well, though, since both buttons will link to the same directory there for some reason.
- Updating causes the following exception after the download is done:
```
Exception in thread Thread-5:
Traceback (most recent call last):
  File "/usr/lib/python3.7/threading.py", line 917, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.7/threading.py", line 865, in run
    self._target(*self._args, **self._kwargs)
  File "/home/wouter/Sources/minigalaxy/minigalaxy/download.py", line 25, in finish
    self.__finish_func()
  File "/home/wouter/Sources/minigalaxy/minigalaxy/ui/gametile.py", line 281, in __update
    install_game(self.game, self.update_path, parent_window=self.parent)
TypeError: install_game() got an unexpected keyword argument 'parent_window'
```

I'd understand if you're no longer motivated to work on this, but if you would, could you please separate this PR into multiple different ones. It's like 10x easier for me to review a single feature instead of a set of 6 different ones. I would help you out with trying to get it to work, but right now this is very hard.
